### PR TITLE
fix(eslint-plugin): [class-literal-property-style] broken fixer when a readonly property is assigned in the constructor

### PR DESCRIPTION
**Pull request by JBhagat841**
A breaking change was reported (#2206 ) such that when the following rules are applied, the following two cases are incorrectly showing errors:

Rules:
```
{
  "rules": {
    "@typescript-eslint/class-literal-property-style": ["error", "getters"],
    "@typescript-eslint/prefer-readonly": "warn"
  }
}
```

```
export class A {
  private readonly foo: string = 'bar';

  constructor(foo: string) {
    this.foo = foo;
  }
}
```
Expected Result
Code should be unchanged

Actual Result
Code after fix is invalid
```
export class A {
  private get foo() { return 'bar'; }

  constructor(foo: string) {
    this.foo = foo;
  }
}
```

Versions applied when bug was reported:

package | version
-- | --
@typescript-eslint/eslint-plugin | 3.2.0
@typescript-eslint/parser | 3.2.0
TypeScript | 3.9.5
ESLint | 7.2.0
node | 14.2.0
npm | 6.14.5
yarn | 1.22.4

### CODE REVIEW

**Comment by JBhagat841** [comment]
@JoshuaKGoldberg throwing up a PR for this and wanted to keep you in the loop.

File: packages/eslint-plugin/tests/rules/class-literal-property-style.test.ts [hunk 1]
```diff
@@ -177,6 +177,24 @@ class Mx {
       `,
       options: ['getters'],
     },
+    {
+      code: `
+class Mx {
+  private readonly foo: string = 'bar';
```
Comment (JBhagat841) [inline packages/eslint-plugin/tests/rules/class-literal-property-style.test.ts:183]: @bradzacher so change this test case to:

```
export class A {
  private readonly foo: string;

  constructor(foo: string) {
    this.foo = foo;
  }
}
```

Comment (bradzacher) [reply packages/eslint-plugin/tests/rules/class-literal-property-style.test.ts:183]: I would have both.

`private readonly foo: string = 'bar';` should be an `invalid` case for the config `options: ['getters']`, and it should be fixed appropriately.

As mentioned; the case with the assignment in the constructor should still be invalid, but there should be no auto-fix for it (use `output: null` to test this).

File: packages/eslint-plugin/src/rules/class-literal-property-style.ts [hunk 2]
```diff
@@ -100,7 +102,31 @@ export default util.createRule<Options, MessageIds>({
     }
 
     return {
+      'MemberExpression > Identifier:exit'(node: TSESTree.Identifier): void {
```
Comment (JBhagat841) [inline packages/eslint-plugin/src/rules/class-literal-property-style.ts:105]: @JoshuaKGoldberg read the documentation you suggested. Do you think I'm getting warmer on having a solution that solves the actual bug?

I realize this code can be much cleaner, but wanted to just know if I'm going in the right direction.

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/class-literal-property-style.ts:105]: Ok cool @JBhagat841 and I discussed this a bit; summarizing the key takeaways I had:
* Rule walkers do a single linear traversal through the node tree, so these `:exit`s don't quote match up with the expected behavior.
* The strategy going forward will be to:
    1. Collect class properties that might need to be _warned_ or _errored_ on as they are found
    2. Inside the class constructor, mark any property assignments as _warnings_ for those class properties
    3. Upon exiting the class constructor, mark any remaining property assignments as _errors_
* As with `prefer-readonly`, you might have to use parser services and will almost certainly need a stack of classes to make sure you're applying node warnings to the right class

That last point is where the trickiness comes in. Please enjoy these very convoluted edge cases 😄 

```ts
const keyName1 = 'wat1'
const keyName2 = 'wat2'


class Parent {
  private readonly wat1 = true;
  private readonly wat2 = true;

  constructor() {
    this['wat1'] = true;
    this[keyName1] = true;
    this.wat1 = true;

    class Child {
      private readonly wat1 = true;

      constructor() {
        this['wat1'] = true;
        this[keyName1] = true;

        console.log(class GrandChild extends Child {
          private readonly wat2 = true;

          constructor() {
            super();
            this.wat2 = true;
            this[keyName2] = true;
          }
        })
      }
    }
  }
}
```

_(personally, I don't think we need to consider absurdities such as `this[keyName1]` as examples of a property being written to for this rule, as those are often impossible to accurately understand the resolved key name in static analysis...)_

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/class-literal-property-style.ts:105]: > you might have to use parser services

Requiring type information is a breaking change, so I'd prefer to not use it unless we reallllyyyy need it.

Comment (JBhagat841) [reply packages/eslint-plugin/src/rules/class-literal-property-style.ts:106]: Translating this into test cases:  

1. test case of computed value this['wat1'] = 123 
2. rescursive case of class in constructor

would the following 2 invalid cases capture what you listed?

```
class Mx {
  private readonly wat1 = true;

  constructor() {
    this['wat1'] = true;
  }
}
```

```
class Mx {
  constructor() {
    class Child {
      private readonly foo = true;

      constructor() {
        this.foo = true;
      }
  	}
  }
}
```

**Pull request closed without merging**